---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-"
  #out.width = "100%"
)
```

# dremeR

<!-- badges: start -->
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![Project Status: WIP â€“ Initial development is in progress, but there has not yet been a stable, usable release suitable for the public.](https://www.repostatus.org/badges/latest/wip.svg)](https://www.repostatus.org/#wip)
<!-- badges: end -->

An R interface to the [MEME Suite](http://meme-suite.org/) family of tools.

## Installation

### Development Version
You can install the development version of dremeR from [GitHub](https://github.com/snystrom/dremeR) with:

```{r, eval=F}
# install.packages("remotes")
remotes::install_github("snystrom/dremeR")
```


## Detecting the MEME Suite

dremeR relies on a local install of the [MEME Suite](http://meme-suite.org/).
For installation instructions for the MEME suite, see the [MEME Suite
Installation Guide](http://meme-suite.org/doc/install.html?man_type=web).

dremeR needs to know the location of the `meme/bin/` directory on your local machine.
You can tell dremeR the location of your MEME suite install in 4 ways. dremeR
will always prefer the more specific definition if it is a valid path. Here they
are ranked from most- to least-specific:

1. Manually passing the install path to the `meme_path` argument of all dremeR functions
2. Setting the path using `options(meme_bin = "/path/to/meme/bin/")` inside your R script
3. Setting `MEME_BIN=/path/to/meme/bin/` in your `.Renviron` file
4. dremeR will try the default MEME install location `~/meme/bin/`

If dremeR fails to detect your install at the specified location, it will fall
back to the next option.

To verify dremeR can detect your MEME install, use `check_meme_install()` which
uses the search herirarchy above to find a valid MEME install. It will report
whether any tools are missing, and print the path to MEME that it sees. This can
be useful for troubleshooting issues with your install.
```{r check_install_works}
library(dremeR)

# Verify that dremeR detects your meme install
# (returns all green checks if so)
# (I have MEME installed to the default location)
check_meme_install()
```

```{r check_install_fails}
# You can manually input a path to meme_path
# If no meme/bin is detected, will return a red X
check_meme_install(meme_path = 'bad/path')
```

## The Core Tools

- `runDreme()`
- `runAme()`
- `runTomTom()`
- `runFimo()`

dremeR is built around the [universalmotif package](https://www.bioconductor.org/packages/release/bioc/html/universalmotif.html)
which provides a framework for manipulating motifs in R.

### Motif Discovery with DREME

```{r}
library(magrittr)
library(GenomicRanges)

# Transcription factor peaks
peaks <- readr::read_tsv(system.file("extdata/peaks/peaks.tsv", package = "dremeR")) %>% 
  GRanges

# Genome object
dm.genome <- BSgenome.Dmelanogaster.UCSC.dm6::BSgenome.Dmelanogaster.UCSC.dm6
```
```{r}
sequences <- peaks %>% 
  resize(200, "center") %>% 
  get_sequence(dm.genome)
```

`runDreme` accepts DNAStringSet or a path to a fasta file as input. You can use
other sequences or shuffled input sequences as the control dataset.
```{r}
# runDreme accepts all arguments that the commandline version of dreme accepts
# here I set e = 50 to detect motifs in the limited example peak list
dreme_results <- runDreme(sequences, control = "shuffle", e = 50)
```
dremeR is built around the [universalmotif](https://www.bioconductor.org/packages/release/bioc/html/universalmotif.html) package. The runDreme results object includes a `motif` column of `universalmotif` objects which can be used in all `universalmotif` functions.
```{r}
library(universalmotif)
view_motifs(dreme_results$motif)
```

### Matching motifs using TOMTOM
Discovered motifs can be matched to known TF motifs using `runTomTom`, which can accept as input a path to a .meme formatted file, a `universalmotif` list, or the results of `runDreme`.

TomTom uses a database of known motifs which can be passed to the `database` parameter as a path to a .meme format file, or a `universalmotif` object.

Optionally, you can set the environment variable `MEME_DB` in `.Renviron`, or
the `meme_db` value in `options` to a valid .meme format file and dremeR will
use that file as the database. dremeR will always prefer user input to the
function call.

```{r}
options(meme_db = system.file("extdata/db/fly_factor_survey_id.meme", package = "dremeR"))
m <- create_motif("CMATTACN", altname = "testMotif")
tomtom_results <- runTomTom(m)
```
```{r}
tomtom_results
```

### Motif Enrichment using AME

```{r}
ame_results <- runAme(sequences, control = "shuffle", evalue_report_threshold = 30)
ame_results
```


## Visualizing Results

```{r}
dreme_results %>% 
  runTomTom() %>% 
  view_tomtom_hits(top_n = 1) %>% 
  cowplot::plot_grid(plotlist = .)
```
```{r}
ame_results %>% 
  ame_plot_heatmap()
```



# FAQs
### How do I use dremeR/MEME on Windows?
The MEME Suite does not currently support Windows, although it can be
installed under [Cygwin](https://www.cygwin.com/) or the [Windows Linux Subsytem](https://docs.microsoft.com/en-us/windows/wsl/install-win10) (WSL).
Please note that if MEME is installed on Cygwin or WSL, you must also run R
inside Cygwin or WSL to use dremeR.
